app/Http/Controllers/InternetController.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\MacAddress;
+use Illuminate\Auth\Access\AuthorizationException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use App\Utils\TabulatorPaginator;
+
+class InternetController extends Controller
+{
+    public function index()
+    {
+        return view('internet.app');
+    }
+
+
+    public function admin()
+    {
+        return view('admin.internet.app');
+    }
+
+    public function getUsersMacAddresses(Request $request)
+    {
+        $paginator = TabulatorPaginator::from(Auth::user()->macAddresses())
+            ->sortable(['mac_address', 'comment', 'state'])->paginate();
+
+        $paginator->getCollection()->transform($this->translateStates());
+
+        return $paginator;
+    }
+
+    public function getUsersMacAddressesAdmin(Request $request)
+    {
+        $paginator = TabulatorPaginator::from(MacAddress::join('users as user', 'user.id', '=', 'user_id')->select('mac_addresses.*')->with('user'))
+            ->sortable(['mac_address', 'comment', 'state', 'user.name'])
+            ->filterable(['mac_address', 'comment', 'user.name', 'state'])
+            ->paginate();
+
+        $paginator->getCollection()->transform($this->translateStates());
+
+        return $paginator;
+    }
+
+    public function deleteMacAddress(Request $request, $id)
+    {
+        $macAddress = MacAddress::findOrFail($id);
+
+        if (!Auth::user()->isAdmin() && $macAddress->user->id != Auth::user()->id) { // TODO: use gate
+            throw new AuthorizationException();
+        }
+
+        $macAddress->delete();
+    }
+
+    public function editMacAddress(Request $request, $id)
+    {
+        if (!Auth::user()->isAdmin()) { // TODO: use gate
+            throw new AuthorizationException();
+        }
+
+        $macAddress = MacAddress::findOrFail($id);
+        if($request->has('state')) {
+            $macAddress->state = $request->input('state');
+        }
+
+        $macAddress->save();
+
+        return $this->translateStates()($macAddress);
+    }
+
+    public function addMacAddress(Request $request)
+    {
+        $request->validate(
```

Comment:
It would be more comfortable to return invalid data with the input. You can use `redirect()->back()->withInput();`.
Or another option would be to add them as fronted checks too.

Reply:
The validate method does exactly this redirection when the validation fails.

Reply:
Are you sure? I think it behaved differently, but I might not remember well. We'll see then later.